docker-unified/entrypoint.sh

```diff
@@ -1,13 +1,5 @@
 #!/bin/bash
 
-# Set up user group
-PGID="${PGID:-1000}"
-addgroup --gid $PGID bitwarden
-
-# Set up user
-PUID="${PUID:-1000}"
```

Comment:
Is this change not forcing the ids to be 1000 ? What if i need a different id to be used for bitwarden to read/write to volumes on the host that aren't 1000 ?

As then breaks other workflows where PGID/PUID could be set to a custom value ensure the right permissions and is very common across docker containers.

Reply:
I see your point, here, let me run some tests.

Reply:
There is a chicken-egg problem here: to run all that "chowns" we need to be a superuser, and we don't want to be a superuser, that's the point of a change. Running container from the normal user is a hardening recommendation from all security frameworks. We cannot add the user to the sudoers with NOPASS, change everything and then remove it from the sudoers because it would violate the next wanted security control - immutable file system in the container which is now also immpossible.

Since Bitwarden is by the nature security related app that keeps all the passwords, secrets and OTP's it would be logical to have it designed in the way it complies with the security frameworks like NSA or MITRE.

I'd checked other projects, forcing ID is common practice.

Reply:
Sure, just ensure that the use case of using docker userns or rootless where root inside the container maps to an unprivileged user outside the container. So then having a further offset of 1000 is not preferable as it could cause permission problems with volumes etc.

If there is still a way to arbitrarily pick or force a user id for the bitwarden process then this should be good still.

Reply:
> userns or rootless where root inside the container maps to an unprivileged user outside the container.

please can you be more specific here, I am not really getting th use case.

You can and most probably should build your own image in your CI if you're running it in prod to set custom UID/GID.

Reply:
If you have configured https://docs.docker.com/engine/security/userns-remap/ or https://docs.docker.com/engine/security/rootless/ or similar with podman, then your container's root on the host is mapped to the subuid of the unprivileged user on the host.

In this case there is less risk of running the process inside the container as root as if it does escape it's actually running as the unprivileged user. So it's commonly used as an extra security layer.

Then volumes that are bind-mounted in (eg to store database or settings etc) you would need to be readable/writable for the bitwarden process, so in these use cases you need to make sure that the permissions inside and outside the container match correctly with the uids. If you can pick the uid for the bitwarden process this makes it remain consistent with the permissions as you can then match it to the file and user permissions on the host correctly, rather than the bitwarden process being forced to an offset of 1000.

I'd rather not have to maintain a whole image that is not part of the bitwarden pipeline, just to have a subtly different UID that most other container images allow configuring.

Reply:
Well, I had checked best practices here: https://docs.docker.com/build/building/best-practices/#user :

> Consider an explicit UID/GID.

Also, best practice is to "[Create ephemeral containers](https://docs.docker.com/build/building/best-practices/#create-ephemeral-containers)", we need to have a container that doesn't perform final configuration touches every time it is spawned.

Currently, the entrypoint.sh performs significant changes on the system on every container start, which is antipattern for the Docker images. As a result, we cannot have immutable container, and we have to run it as superuser. These two are common controls in all security frameworks I heard about, and Bitwarden is to be the first hardened application in the infra.

> Then volumes that are bind-mounted in (eg to store database or settings etc) you would need to be readable/writable for the bitwarden process, so in these use cases you need to make sure that the permissions inside and outside the container match correctly with the uids. 

I cannot really imagine a situation when it would be an issue, other than the data volumes were created with strange UID's for some reason.

> I'd rather not have to maintain a whole image that is not part of the bitwarden pipeline, just to have a subtly different UID that most other container images allow configuring.

Sorry, not getting this part, what container images allow this? Docker best practices clearly highlight this as the antipattern.

Reply:
Some examples of using PUID etc are from the linuxserver.io images that have standardised around that https://docs.linuxserver.io/general/understanding-puid-and-pgid/

But I also note that there is the `docker run --user` flag, maybe this can be made to work so that the user can be UID=1000 in the images but have the correct file permissions on the files in the image so that developers can override the user uid with `--user 123` ?

One usecase where this is important would be running backups of volumes from another container. If you have already chosen that your backups are running as id 123 then you want bitwarden to also be creating it's config / running under id 123 so that the backup task can access the files.

So maybe the summary is, can the files that your are chown'ing in image `/app /etc /var/ /run` etc ensure that they have permissions to allow for developers to override the UID to something else? :thinking: And then work with the docker user override? https://docs.docker.com/engine/reference/run/#user

Reply:
But, i should be able to use bindfs ( https://bindfs.org/ ) to create mounts that maps ids between X and Y, so if this does land i can then map the permissions across to the expected id.